dsc/workflows/base/__init__.py

```diff
@@ -0,0 +1,207 @@
+        email_recipients: list[str],
+        metadata_mapping: dict,
+        s3_bucket: str,
+        s3_prefix: str | None,
+        collection_handle: str,
+        output_queue: str,
+    ) -> None:
+        """Initialize base instance.
+
+        Args:
+            workflow_name: The name of the workflow.
+            submission_system: The system to which item submissions will be sent (e.g.
+            DSpace@MIT)
+            email_recipients: The email addresses to notify after runs of the workflow.
+            metadata_mapping: A mapping file for generating DSpace metadata from the
+            workflow's source metadata.
+            s3_bucket: The S3 bucket containing bitstream and metadata files for the
+            workflow.
+            s3_prefix: The S3 prefix used for objects in this workflow. This prefix does
+            NOT include the bucket name.
+            collection_handle: The handle of the DSpace collection to which submissions
+            will be uploaded
+            output_queue: The SQS output queue used for retrieving result messages from
+            the workflow's submissions.
+        """
+        self.workflow_name: str = workflow_name
+        self.submission_system: str = submission_system
+        self.email_recipients: list[str] = email_recipients
+        self.metadata_mapping: dict = metadata_mapping
+        self.s3_bucket: str = s3_bucket
+        self.s3_prefix: str | None = s3_prefix
+        self.collection_handle: str = collection_handle
+        self.output_queue: str = output_queue
+
+    @final  # noqa: B027
+    def run(self) -> None:
+        """Run workflow to submit items to  the DSpace Submission Service.
+
+        PLANNED CODE:
+
+        sqs_client = SQSClient()
+        for item_submission in self.item_submissions_iter():
+            item_submission.upload_dspace_metadata(
+                self.s3_bucket, item_submission.metadata_s3_key
+            )
+            sqs_client.send_submission_message(
+                item_submission.item_identifier,
+                self.workflow_name,
+                self.output_queue,
+                self.submission_system,
+                self.collection_handle,
+                item_submission.metadata_uri,
+                item_submission.bitstream_uris,
+            )
+        """
+
+    @final
+    def item_submissions_iter(self) -> Iterator[ItemSubmission]:
+        """Generate a batch of item submissions for the DSpace Submission Service.
+
+        MUST NOT be overridden by workflow subclasses.
+        """
+        for item_metadata in self.batch_metadata_iter():
+            item_identifier = self.get_item_identifier(item_metadata)
+            logger.info(f"Processing submission for '{item_identifier}'")
+            metadata_s3_key = f"{self.s3_prefix}/{item_identifier}_metadata.json"
+            dspace_metadata = self.create_dspace_metadata(item_metadata)
+            self.validate_dspace_metadata(dspace_metadata)
```

Comment:
Looking at this now, I could envision validating the generated metadata _inside_ the creation.  But I think there are arguably advantages to doing it this way as well.  So NOT a request, just an observation, in case the pattern emerges elsewhere.